File: SearchInRotated.py
# Time Complexity: O(logn)
# Space Complexity: O(1)
# Did this code successfully run on Leetcode: Yes
# Any problem you faced while coding this: No

class SearchInRotated:

    def Search(self, arr, target):
        low = 0
        high = len(arr)-1
        while low <= high:
            mid = low + (high-low)//2
            if arr[mid] == target:
                return mid
            # left is sorted, check if target is in this range. then move high to mid-1
            elif arr[low] <= arr[mid]:
                if arr[low] <= target < arr[mid]:
                    high = mid - 1
                else:
                    low = mid + 1
            else:
                if arr[mid] < target <= arr[high]:
                    low = mid + 1
                else:
                    high = mid - 1
        return -1


obj = SearchInRotated()
arr = [3, 1]
target = 3
print(obj.Search(arr, target))

File: SearchInRotated.cpp
// Time Complexity : O(log(n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
#include <iostream>
#include <vector>

using namespace std;
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int i = 0;
        int j = nums.size()-1;
        while(i<=j) {
            int mid = (j-i)/2 +i;
            if (nums[mid] == target) return mid;
            else if (nums[i] <= nums[mid]) {
                if (nums[i] <= target && target < nums[mid]) j = mid -1;
                else i = mid+1;
            }
            else{
                if (target <= nums[j] && nums[mid] < target) i = mid+1;
                else j = mid-1;
            }
        }
        return -1;
    }
};